### PR TITLE
#162948806 get one article

### PR DESCRIPTION
**Pull request by kwameasiago**
#### What does this PR do?
Enable users to fetch single article via a slug

#### Description of Task to be completed?
* [x]   Create actions for fetching a single article
* [x]   Create reducer for fetching a single article
* [x]   Consume the `get  single article` API endpoint
* [x]   Created test for the front end

#### How should this be manually tested?
1. Clone this repository on your local machine using
   `git clone https://github.com/andela/ah-django-unchained-frontend.git`
2. Change your working directory to the project's directory using
   `cd ah-django-unchained-frontend`
3. Change to the specific branch `ft-get-one-article-1629488 ` using
   `git checkout ft-get-one-article-16294880 `
4. Create a `.env` file and inside it add this line `REACT_APP_BASE_URL=app-base-url` where `app-base-url ` is the root url of the backend API
5. Run `npm install` to install all the modules listed in the `package.json` file.
6. Run `npm start` command to start the application.
7. The application should start successfully. On the browser, navigate to the URL `http://localhost:3000/article/artilce-slug`, .

#### Link to PT story
[#162948806](https://www.pivotaltracker.com/story/show/162948806)

#### Screenshots
screenshot 2019-02-27 at 22 40 47

screenshot 2019-02-27 at 22 41 00

### CODE REVIEW

File: src/store/modules/articleDetail/index.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+    default:
+      return{
+        ...state,
+        isFetch:false
+      };
+  }
+};
+
+export default singleArticleReducer;
```
Comment (Georgeygigz) [inline src/store/modules/articleDetail/index.js]: kindly add a blank line at the end of the file

File: src/store/modules/articleDetail/type.js [hunk 2]
```diff
@@ -0,0 +1,2 @@
+export const ARTICLE_SUCCESS = 'ARTICLE_SUCCESS';
+export const ARTICLE_FAIL = 'ARTICLE_FAIL';
```
Comment (Georgeygigz) [inline src/store/modules/articleDetail/type.js]: ditto

File: src/store/modules/articleDetail/articleDetail.reduce.test.jsx [hunk 3]
```diff
@@ -0,0 +1,37 @@
+import {singleArticleReducer} from './index';
+import {ARTICLE_SUCCESS, ARTICLE_FAIL} from './type';
+
+describe('sign up reducer', () => {
+  it('should run with  the initial state and return isFetch false', () =>{
+    const expectedData = {
+    isFetch: false
+    };
```
Comment (MaggieChege) [inline src/store/modules/articleDetail/articleDetail.reduce.test.jsx]: Rename file to articledDetail.reducers.test.jsx

File: src/store/modules/articleDetail/articleDetail.actions.test.jsx [hunk 4]
```diff
@@ -0,0 +1,68 @@
+import moxios from 'moxios';
+import {ARTICLE_SUCCESS, ARTICLE_FAIL} from './type';
+import { mockStore } from '../../../utils/helpers/testHelpers';
+import {getSingleArticleDispatch} from './index';
+
+describe('ArticleSingle', () => {
```
Comment (andrewhingah) [inline src/store/modules/articleDetail/articleDetail.actions.test.jsx]: The title of your describe block should be the name of the component you are testing, in this case `ArticleDetail`

File: src/store/modules/articleDetail/articleDetail.reducers.test.jsx [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+describe('sign up reducer', () => {
+  it('should run with  the initial state and return isFetch false', () =>{
+    const expectedData = {
+    isFetch: false
+    };
+    const action = {
+      type: undefined
+    };
+
+    expect(singleArticleReducer({},action)).toEqual(expectedData);
```
Comment (andrewhingah) [inline src/store/modules/articleDetail/articleDetail.reducers.test.jsx]: Please remove this blank line

File: src/store/modules/articleDetail/articleDetail.reducers.test.jsx [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  it('should run with ARTICLE SUCCESS return isFetch true', () =>{
+    const expectedData = {
+    isFetch: true,
+    data: undefined,
+    };
+    const action = {
+      type: ARTICLE_SUCCESS
+    };
+
+    expect(singleArticleReducer({},action)).toEqual(expectedData);
```
Comment (andrewhingah) [inline src/store/modules/articleDetail/articleDetail.reducers.test.jsx]: ditto

File: src/store/modules/articleDetail/articleDetail.reducers.test.jsx [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  it('should run with ARTICLE FAIL return isFetch true', () => {
+    const expectedData = {
+    isFetch: false,
+    data: undefined,
+    };
+    const action = {
+      type: ARTICLE_FAIL
+    };
+    
+    expect(singleArticleReducer({},action)).toEqual(expectedData);
```
Comment (andrewhingah) [inline src/store/modules/articleDetail/articleDetail.reducers.test.jsx]: ditto

Comment (kwameasiago) [reply src/store/modules/articleDetail/articleDetail.reducers.test.jsx]: I put that line there for readability

**Comment by cosmas28** [comment]
Kindly Fix merge conflicts.

File: src/store/modules/articleDetail/type.js [hunk 8]
```diff
@@ -0,0 +1,2 @@
+export const ARTICLE_SUCCESS = 'ARTICLE_SUCCESS';
```
Comment (cosmas28) [inline src/store/modules/articleDetail/type.js]: This action types should more descriptive, ie. for instance, `GET_SINGLE_ARTICLE_SUCCESS`

**Review by cosmas28 — CHANGES_REQUESTED** [review]

File: src/components/NotFound/NotFound.test.jsx [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+import {shallow} from 'enzyme';
+import {NotFound} from './index';
+
+describe('NotFound', ()=>{
```
Comment (KMaina) [inline src/components/NotFound/NotFound.test.jsx]: Maybe you can have spaces after the parenthesis and opening brace

File: src/store/modules/articleDetail/articleDetail.actions.test.jsx [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    moxios.install();
+  });
+
+  afterEach(() => {
+    moxios.uninstall();
+  });
+
+  it('should fetch single articles with success', () => {
+    const mockHttpResponse = {
+      status: 201,
```
Comment (KMaina) [inline src/store/modules/articleDetail/articleDetail.actions.test.jsx]: fetching an article should have a status code of 200

File: src/store/modules/articleDetail/articleDetail.actions.test.jsx [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+      data: { status: 201 }
```
Comment (KMaina) [inline src/store/modules/articleDetail/articleDetail.actions.test.jsx]: kindly look at your status code and amend it

File: src/store/modules/articleDetail/articleDetail.reducers.test.jsx [hunk 12]
```diff
@@ -0,0 +1,39 @@
+import {singleArticleReducer} from './index';
+import {GET_SINGLE_ARTICLE_SUCCESS, GET_SINGLE_ARTICLE_FAIL} from './type';
+
+describe('sign up reducer', () => {
+  it('should run with  the initial state and return isFetch false', () =>{
+    const expectedData = {
+    isFetch: false
```
Comment (KMaina) [inline src/store/modules/articleDetail/articleDetail.reducers.test.jsx]: indent this line here

File: src/store/modules/articleDetail/articleDetail.reducers.test.jsx [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+    const action = {
+      type: undefined
+    };
+
+    expect(singleArticleReducer({},action)).toEqual(expectedData);
+  });
+
+  it('should run with ARTICLE SUCCESS return isFetch true', () =>{
+    const expectedData = {
+    isFetch: true,
```
Comment (KMaina) [inline src/store/modules/articleDetail/articleDetail.reducers.test.jsx]: indent code within this block

File: src/store/modules/articleDetail/articleDetail.reducers.test.jsx [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+    const action = {
+      type: GET_SINGLE_ARTICLE_SUCCESS
+    };
+
+    expect(singleArticleReducer({},action)).toEqual(expectedData);
+  });
+
+  it('should run with ARTICLE FAIL return isFetch true', () => {
+    const expectedData = {
+    isFetch: false,
```
Comment (KMaina) [inline src/store/modules/articleDetail/articleDetail.reducers.test.jsx]: indent this line

File: src/pages/ArticleDetail/ArticDetail.test.jsx [hunk 15]
```diff
@@ -1,15 +1,56 @@
...
+  
   beforeEach(() => {
-    wrapper = shallow(<ArticleDetail />);
+    props = {
+      data: {
+        title: 'title',
+        description: 'description',
+        body: 'body'
+      },
+      IsFetch:false,
```
Comment (KMaina) [inline src/pages/ArticleDetail/ArticDetail.test.jsx]: kindly add a space after `IsFetch` to make your code meet linting standards

File: src/pages/ArticleDetail/ArticDetail.test.jsx [hunk 16]
```diff
@@ -1,15 +1,56 @@
...
+      match: {
+        params: {
+          article: 'article'
+        }
+      }
+    };
+
+    state = {
+      data: {},
+      IsFetch:false
```
Comment (KMaina) [inline src/pages/ArticleDetail/ArticDetail.test.jsx]: kindly add a space after `IsFetch` to make your code meet linting standards

File: src/store/modules/articleDetail/articleDetail.reducers.test.jsx [hunk 17]
```diff
@@ -0,0 +1,39 @@
+import { singleArticleReducer } from './index';
+import { GET_SINGLE_ARTICLE_SUCCESS, GET_SINGLE_ARTICLE_FAIL } from './type';
+
+describe('sign up reducer', () => {
+  it('should run with  the initial state and return isFetch false', () => {
+    const expectedData = {
+      isFetch: false
+    };
+    const action = {
+      type: undefined
```
Comment (d-kahara) [inline src/store/modules/articleDetail/articleDetail.reducers.test.jsx:11]: Any reason why this action type is declared as `undefined` ?

File: src/store/modules/articleDetail/index.js [hunk 18]
```diff
@@ -0,0 +1,54 @@
+import  {GET_SINGLE_ARTICLE_SUCCESS, GET_SINGLE_ARTICLE_FAIL} from './type';
+import http from '../../../utils/helpers/http';
+
+//  Actions for getting single profle
+const getSingleArticleSuccess = payload => ({
+  type: GET_SINGLE_ARTICLE_SUCCESS,
+  payload
+});
+
+const  failSingleArticleFail = payload  => ({
```
Comment (cosmas28) [inline src/store/modules/articleDetail/index.js]: The name of this action creator is not descriptive enough, kindly sync with your teammates to refactor all the action names.

File: src/store/modules/articleDetail/articleDetail.reducers.test.jsx (hunk 17 above)
Comment (kwameasiago) [reply src/store/modules/articleDetail/articleDetail.reducers.test.jsx:11]: am testing initial state that uses the default action

File: src/store/modules/articleDetail/index.js [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+    .then((res) => {
+      dispatch(getSingleArticleSuccess(res.data));
+    })
+    .catch((error) => {
+      dispatch(failSingleArticleFail(error));
+    });
+};
+
+// reducers for articles
+export const singleArticleReducer = (state = {data: {}}, action) =>{
```
Comment (cosmas28) [inline src/store/modules/articleDetail/index.js]: The `Reducer` prefix in this reducer is not necessary, kindly remove it.

**Review by cosmas28 — CHANGES_REQUESTED** [review]

**Comment by cosmas28** [comment]
Kindly note the following:
- rebase from developing and fix the conflicts.
- add proper navigation for a single article from the Article list page.

File: src/store/modules/articleDetail/articleDetail.reducers.test.jsx [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+      isFound: true
+    };
+    const action = {
+      type: GET_SINGLE_ARTICLE_SUCCESS
+    };
+
+    expect(singleArticleReducer({}, action)).toEqual(expectedData);
+  });
+
+  it('should run with ARTICLE FAIL return isFetch true', () => {
```
Comment (andrewhingah) [inline src/store/modules/articleDetail/articleDetail.reducers.test.jsx]: The title of this `it` block is confusing. The isFetch returns false, and not true as stated.

File: src/store/modules/articleDetail/articleDetail.reducers.test.jsx (hunk 17 above)
Comment (andrewhingah) [reply src/store/modules/articleDetail/articleDetail.reducers.test.jsx:11]: It doesn't return undefined when I ran locally. 
``` 
Expected value to equal:
{"data": undefined, "isFetch": true}
Received:
{"isFetch": false, "isFound": true}
```

Comment (kwameasiago) [reply src/store/modules/articleDetail/articleDetail.reducers.test.jsx:11]: I have made changes please pull and try again

File: src/store/modules/articleDetail/index.js (hunk 19 above)
Comment (cosmas28) [reply src/store/modules/articleDetail/index.js]: Any reason why this feedback has not been implemented?

**Review by cosmas28 — CHANGES_REQUESTED** [review]

File: src/store/modules/articleDetail/index.js (hunk 19 above)
Comment (kwameasiago) [reply src/store/modules/articleDetail/index.js]: I have fixed it

File: src/store/modules/articleDetail/type.js [hunk 21]
```diff
@@ -0,0 +1,3 @@
+export const GET_SINGLE_ARTICLE_SUCCESS = 'GET_SINGLE_ARTICLE_SUCCESS';
+export const GET_SINGLE_ARTICLE_FAIL = 'GET_SINGLE_ARTICLE_FAIL';
+export const GET_SINGLE_ARTICLE_REQUEST = 'GET_SINGLE_ARTICLE_REQUEST';
```
Comment (MaggieChege) [inline src/store/modules/articleDetail/type.js]: Add Blank line at the end of this file

File: src/pages/ArticleDetail/index.jsx [hunk 22]
```diff
@@ -1,53 +1,54 @@
...
+    getSingleArticle(match.params.article);
+  }
+
+  render() {
+    const { singleArticle } = this.props;
+    const { singleArticle: { isFetching } } = this.props;
+    return (
+      <React.Fragment>
+        {isFetching === true ? <Loader /> :
+
```
Comment (KMaina) [inline src/pages/ArticleDetail/index.jsx]: Kindly remove this space if it is not necessary

**Comment by d-kahara** [comment]
On Instruction no.7 of the PR, shouldn't one navigate to 
 `http://localhost:3000/article/article-slug`?

The Pivotal Tracker Story ID used in the PR header and provided as a link belongs to `Users should be able to create articles`

**Review by cosmas28 — APPROVED** [review]

**Pull request merged**